### PR TITLE
extract all files in target require filedirs

### PR DESCRIPTION
**Pull request by tomoyamachi**
fix for https://github.com/aquasecurity/trivy/issues/96 (https://github.com/aquasecurity/fanal/issues/28)

```
$ go run cmd/fanal/main.go gcr.io/distroless/java:11
{Name:9.9 Family:debian}
via image Packages: 19
via file Packages: 0

$ go run cmd/fanal/main.go debian:latest
{Name:10.0 Family:debian}
via image Packages: 91
via file Packages: 0
```

### CODE REVIEW

File: analyzer/pkg/dpkg/dpkg.go [hunk 1]
```diff
@@ -29,6 +30,19 @@ type debianPkgAnalyzer struct{}
 func (a debianPkgAnalyzer) Analyze(fileMap extractor.FileMap) (pkgs []analyzer.Package, err error) {
 	detected := false
 	for _, filename := range a.RequiredFiles() {
+		// find by target directory
+		if filename[len(filename)-1:] == "/" {
```
Comment (knqyf263) [inline analyzer/pkg/dpkg/dpkg.go]:
```suggestion
		if filename[len(filename)-1] == '/' {
```

File: extractor/docker/docker.go [hunk 2]
```diff
@@ -336,6 +336,14 @@ func (d DockerExtractor) ExtractFiles(layer io.Reader, filenames []string) (extr
 		// Determine if we should extract the element
 		extract := false
 		for _, s := range filenames {
+			// extract all files in target directory if last char is "/"(Separator)
+			if s[len(s)-1:] == "/" {
```
Comment (knqyf263) [inline extractor/docker/docker.go]: Ditto

File: extractor/docker/docker.go [hunk 3]
```diff
@@ -336,6 +336,14 @@ func (d DockerExtractor) ExtractFiles(layer io.Reader, filenames []string) (extr
...
+				if filepath.Dir(s) == filepath.Dir(filePath) {
```
Comment (knqyf263) [inline extractor/docker/docker.go]: Is filepath.Dir(s) necessary? Already checked.

```suggestion
				if s == filepath.Dir(filePath) {
```

File: analyzer/pkg/dpkg/dpkg.go [hunk 4]
```diff
@@ -29,6 +30,19 @@ type debianPkgAnalyzer struct{}
...
+			for targetname, targetbytes := range fileMap {
+				if filepath.Dir(filename) == filepath.Dir(targetname) {
```
Comment (knqyf263) [inline analyzer/pkg/dpkg/dpkg.go]: Ditto

File: analyzer/pkg/dpkg/dpkg.go [hunk 5]
```diff
@@ -144,5 +158,5 @@ func (a debianPkgAnalyzer) parseDpkgPkg(scanner *bufio.Scanner) (pkg *analyzer.P
 }
 
 func (a debianPkgAnalyzer) RequiredFiles() []string {
-	return []string{"var/lib/dpkg/status"}
+	return []string{"var/lib/dpkg/status", "var/lib/dpkg/status.d/"}
```
Comment (knqyf263) [inline analyzer/pkg/dpkg/dpkg.go]:
```suggestion
	return append(a.requiredFiles(), a.requiredDirs()...)

func (a debianPkgAnalyzer) requiredFiles() []string {
	return []string{"var/lib/dpkg/status"}
}	
func (a debianPkgAnalyzer) requiredDirs() []string {
	return []string{"var/lib/dpkg/status.d/"}
}
```

If you care about a performance, we can define these variables in advance.

File: analyzer/pkg/dpkg/dpkg.go [hunk 6]
```diff
@@ -29,6 +30,19 @@ type debianPkgAnalyzer struct{}
 func (a debianPkgAnalyzer) Analyze(fileMap extractor.FileMap) (pkgs []analyzer.Package, err error) {
 	detected := false
 	for _, filename := range a.RequiredFiles() {
+		// find by target directory
+		if filename[len(filename)-1:] == "/" {
+			for targetname, targetbytes := range fileMap {
```
Comment (knqyf263) [inline analyzer/pkg/dpkg/dpkg.go]: It seems better to make this loop the outer.

Before
```
for _, filename := range a.RequiredFiles()
    for targetname, targetbytes := range fileMap
```

After
```
for targetName, targetBytes := range fileMap
    dir := filepath.Dir(targetName)
    if !utils.StringInSlice(targetName, a.requiredFiles()) && !utils.StringInSlice(dir, a.requiredDirs()){
        continue
    }
```

File: extractor/docker/docker.go (hunk 3 above)
Comment (tomoyamachi) [reply extractor/docker/docker.go]: s : `var/lib/dpkg/status.d/`
filepath.Dir(s) : `var/lib/dpkg/status.d`

Comment (tomoyamachi) [reply extractor/docker/docker.go]: `s[0:len(s)-2] == filepath.Dir(filePath)` looks mess.

Comment (knqyf263) [reply extractor/docker/docker.go]: I got it. However, we already know `s` is a directory. It seems a little weird to apply filepath.Dir() to `s`. We may need a comment.

Comment (knqyf263) [reply extractor/docker/docker.go]: How about `filepath.Clean`?

Comment (knqyf263) [reply extractor/docker/docker.go]: >Dir calls Clean on the path and trailing slashes are removed.
https://golang.org/pkg/path/filepath/#Dir

Comment (knqyf263) [reply extractor/docker/docker.go]: We must remove trailing slashes rather than extracting directory name. I wanna make it obvious.

Comment (tomoyamachi) [reply extractor/docker/docker.go]: I fixed it.
I can change OSAnalyzer interface to following if you want.

```
type OSAnalyzer interface {
	Analyze(extractor.FileMap) (OS, error)
	RequiredFiles() []string
	RequiredDirs() []string
}
```

File: analyzer/pkg/dpkg/dpkg.go (hunk 6 above)
Comment (tomoyamachi) [reply analyzer/pkg/dpkg/dpkg.go]: Currently, I added "/" to last char.

```
dir := filepath.Dir(filename) + "/"
```

If you don't like, I can create RequiredDir func.
check => https://github.com/aquasecurity/fanal/pull/29#discussion_r320302883

**Comment by knqyf263** [comment]
Thank you for applying my comments. We need some unittests 😇 
Add image5.tar and directory to filenames:
https://github.com/aquasecurity/fanal/blob/master/extractor/docker/docker_test.go

**Pull request merged**